Lib/fontbakery/profiles/gdef.py

```diff
@@ -0,0 +1,153 @@
+from fontbakery.callable import check
+from fontbakery.checkrunner import FAIL, PASS, WARN
+from fontbakery.message import Message
+from fontbakery.utils import pretty_print_list
+
+# used to inform get_module_profile whether and how to create a profile
+from fontbakery.fonts_profile import profile_factory # NOQA pylint: disable=unused-import
+
+from fontTools import unicodedata
+from fontbakery.constants import GDEF_MARK_GLYPH_CLASS
+
+
+def _is_non_mark_char(charcode):
+  category = unicodedata.category(chr(charcode))
+  if category.startswith("C"):
+    # skip control characters
+    return None
+  elif not category.startswith("M"):
+    return True
+  return False
+
+
+def _get_mark_class_glyphs(ttFont):
+  class_defs = ttFont["GDEF"].table.GlyphClassDef.classDefs.items()
+  return {name for (name, value) in class_defs
+          if value == GDEF_MARK_GLYPH_CLASS}
+
+
+@ check(
+  id = 'com.google.fonts/check/gdef_spacing_marks',
+  rationale = """
+    Glyphs in the GDEF mark glyph class should be non-spacing.
+    Spacing glyphs in the GDEF mark glyph class may have incorrect
+    anchor positioning that was only intended for building composite glyphs
+    during design.
+  """
+)
+def com_google_fonts_check_gdef_spacing_marks(ttFont):
+  """Check mark characters are in GDEF mark glyph class)"""
+  if "GDEF" in ttFont and ttFont["GDEF"].table.GlyphClassDef:
+    spacing_glyphs = {name for (name, value) in ttFont["hmtx"].metrics.items()
+                      if value}
+    mark_class_glyphs = _get_mark_class_glyphs(ttFont)
+    spacing_glyphs_in_mark_glyph_class = spacing_glyphs & mark_class_glyphs
+    if spacing_glyphs_in_mark_glyph_class :
+      formatted_list = "\t " +\
+        pretty_print_list(sorted(spacing_glyphs_in_mark_glyph_class),
+                          shorten=10,
+                          sep=", ")
+      yield WARN,\
+        Message('spacing-mark-glyphs',
+        f"The following spacing glyphs may be in the GDEF mark glyph "
+        f"class by mistake:\n"
+        f"{formatted_list}")
+    else:
+      yield PASS,\
+        'Font does not has spacing glyphs in the GDEF mark glyph class.'
+  else:
+    yield (PASS,
+      'Font does not declare an optional "GDEF" table or has '
+      'any GDEF glyph class definition.')
+
+
+@check(
+  id = 'com.google.fonts/check/gdef_mark_chars',
+  rationale = """
+    Mark characters should be in the GDEF mark glyph class.
+  """,
+)
+def com_google_fonts_check_gdef_mark_chars(ttFont):
+  """Check mark characters are in GDEF mark glyph class"""
+
+  if "GDEF" in ttFont and ttFont["GDEF"].table.GlyphClassDef:
+    cmap = ttFont.getBestCmap()
+    mark_char_glyphs = {
+      charcode for charcode in cmap
+      if _is_non_mark_char(charcode) is False
+    }
+    mark_class_glyphs = _get_mark_class_glyphs(ttFont)
+    mark_char_glyphs_not_in_mark_class = mark_char_glyphs - mark_class_glyphs
```

Comment:
I’m not sure how this code would work, `mark_char_glyphs` is actually a set of code points, and `mark_class_glyphs` is a set of glyph names.

Reply:
(I’m now seeing lots of characters that are reported as missing mark class, while the glyphs really have them).